src-tauri/src/blobs.rs

```diff
@@ -0,0 +1,145 @@
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::{bail, Result};
+use iroh_io::AsyncSliceReaderExt;
+use p2panda_core::Hash;
+use tauri::http::{header, Request, Response, StatusCode, Uri};
+use tauri::{Manager, Runtime, UriSchemeContext, UriSchemeResponder};
+use tokio::sync::RwLock;
+use tokio::task::LocalSet;
+use tokio::time::timeout;
+
+use crate::app::{Context, Rpc};
+
+/// Time limit we give for finding another peer and downloading the blob from them before we give
+/// up and return a "Not found" (404) error.
+const SYNC_TIMEOUT: Duration = Duration::from_secs(5);
+
+/// Handler for the custom blobstore:// URI scheme protocol which will be registered in the Tauri
+/// WebView.
+///
+/// This allows the frontend to request a file addressed with it's hash. First we check if the file
+/// exists in our local blob store, if not, we'll request to download the file from other peers.
+/// This might take a while as no peer with that file might be online, the user will receive a
+/// timeout after a while which can be interpreted as a "not found" 404 failure.
+///
+/// To make this work in Tauri we need to allow this protocol in the CSP config, for example via:
+///
+/// ```json
+/// "csp": {
+///   "img-src": "'self' data: blobstore:"
+/// }
+/// ```
+pub fn blobstore_protocol<R: Runtime>(
+    ctx: UriSchemeContext<'_, R>,
+    request: Request<Vec<u8>>,
+    responder: UriSchemeResponder,
+) {
+    let context = { ctx.app_handle().state::<Rpc>().context.clone() };
+
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    std::thread::spawn(move || {
+        let local = LocalSet::new();
```

Comment:
iroh io methods require to be `!Send`, this is why I'm using the local set here.